src/librustc_lint/non_ascii_idents.rs

```diff
@@ -7,17 +7,37 @@ declare_lint! {
     "detects non-ASCII identifiers"
 }
 
-declare_lint_pass!(NonAsciiIdents => [NON_ASCII_IDENTS]);
+declare_lint! {
+    pub UNCOMMON_CODEPOINTS,
+    Warn,
+    "detects uncommon Unicode codepoints in identifiers"
+}
+
+declare_lint_pass!(NonAsciiIdents => [NON_ASCII_IDENTS, UNCOMMON_CODEPOINTS]);
 
 impl EarlyLintPass for NonAsciiIdents {
     fn check_ident(&mut self, cx: &EarlyContext<'_>, ident: ast::Ident) {
-        if !ident.name.as_str().is_ascii() {
+        use unicode_security::GeneralSecurityProfile;
+        let name_str = ident.name.as_str();
+        if name_str == "$crate" || name_str == "{{root}}" {
+            return;
+        }
+        if !name_str.is_ascii() {
             cx.struct_span_lint(
                 NON_ASCII_IDENTS,
                 ident.span,
                 "identifier contains non-ASCII characters",
             )
             .emit();
         }
+
+        if !name_str.chars().all(GeneralSecurityProfile::identifier_allowed) {
```

Comment:
This check (and all other checks for this rfc) should be inside the non-ascii check.

That will thus not impact performance for ascii, and will also obviate the need for the `$crate` check.

Reply:
Done.